### ISSUE TITLE
serializing Data with lsid=None causes exception on save

### ISSUE THREAD

**Issue by anthonycor** [LabKey/labkey-api-python#5]
When Data objects are added to a Run's inputs or outputs with no lsid (but with a pipelinePath or DataFileUrl) you receive the following java Exception from the server:

`org.labkey.api.view.NotFoundException: Could not find data with LSID null
        at org.labkey.api.data.ExpDataFileConverter.resolveExpData(ExpDataFileConverter.java:80)
        at org.labkey.api.study.assay.DefaultAssaySaveHandler.handleData(DefaultAssaySaveHandler.java:602)
        at org.labkey.api.study.assay.DefaultAssaySaveHandler.getInputData(DefaultAssaySaveHandler.java:350)
        at org.labkey.api.study.assay.DefaultAssaySaveHandler.handleProtocolApplications(DefaultAssaySaveHandler.java:389)
        at org.labkey.api.study.assay.DefaultAssaySaveHandler.handleRun(DefaultAssaySaveHandler.java:284)
        at org.labkey.api.study.assay.DefaultAssaySaveHandler.handleBatch(DefaultAssaySaveHandler.java:144)`

I believe because the ExpObject.to_json is adding the lsid property to the dict to be serialized even when it is None, thus the key exists on the JSONObject [HashMap] server side and then tries to resolve the ExpData object via a null LSID instead of creating a new ExpData from the pipelinePath or DataFileURL.

**Comment by labkey-nicka** [comment]
If you still have a repro of the problem locally, is `LSID` null on the server or is something else (e.g. "null"). The place I'd like to check is `ExpDataFileConverter.java` where the code is:

```java
// Then try as an LSID
else if (dataObject.has(ExperimentJSONConverter.LSID)) // breakpoint here for value in dataObject
{
    ...
}
```

This is likely appropriately fixed on the server similar to `ExperimentJSONConverter.PIPELINE_PATH` handling in that same file.

```java
else if (dataObject.has(ExperimentJSONConverter.LSID) && dataObject.getString(ExperimentJSONConverter.LSID) != null)
{
    ...
}
```

**Comment by anthonycor** [comment]
I agree that the server side null check to match the other JSON property checks would be an appropriate fix here. 

The thought here was to have consistent behavior with the [LabKey JS API for saveBatch](https://www.labkey.org/download/clientapi_docs/javascript-api/symbols/LABKEY.Experiment.html#.saveBatch)

> Saves a modified batch. Runs within the batch may refer to existing data and material objects, either inputs or outputs, by ID or LSID. Runs may also define new data and materials objects by not specifying an ID or LSID in their properties.

**Comment by labkey-nicka** [comment]
Additionally, I've committed the fix I mentioned above with r51799. This change will be available in LabKey 17.2.
